### ISSUE TITLE
Filter doesn't work with nested objects (Firestore maps)

### ISSUE THREAD

**Issue by dhstack** [benwinding/react-admin-firebase#75]
react-admin has the ability to use a reference to a Firestore map by using something like:
`<TextField source="author.firstName" />` (as briefly documented in [the docs](https://marmelab.com/react-admin/Fields.html)).
or
`<ReferenceField label="Organization" source="organization.id" reference="organizations">`

When trying to do this with a filter, it doesn't work.
```
const ProgramFilter = (props) => (
  <Filter {...props}>
    <ReferenceInput label="Organization" source="organization.id" reference="organizations" allowEmpty>
      <SelectInput optionText="name" />
    </ReferenceInput>
  </Filter>
);
```

This is the URL that's generated when the above filter is applied:
```
http://localhost:3000/programs?filter=%7B%22organization%22%3A%7B%22id%22%3A%22ZxaE6XPDbMQWkeADrUYs%22%7D%7D&order=ASC&page=1&perPage=10&sort=id
```
And the [decoded](https://meyerweb.com/eric/tools/dencoder/) version:
```
http://localhost:3000/programs?filter={"organization":{"id":"ZxaE6XPDbMQWkeADrUYs"}}&order=ASC&page=1&perPage=10&sort=id
```

**Comment by dhstack** [comment]
This also causes an issue with the `ReferenceInput` component:

```
    <ReferenceInput
      label="Program"
      source="program.id"
      reference="programs"
      filter={{ organization: 'organization.id' }}
    >
      <SelectInput optionText="name" />
    </ReferenceInput>
```

It *does*, however work with non-nested objects:
```
    <ReferenceInput
      label="Program"
      source="program.id"
      reference="programs"
      filter={{ state: 'Draft' }}
    >
      <SelectInput optionText="name" />
    </ReferenceInput>
```

**Comment by dhstack** [comment]
I figured out a work around based on the example in that @LaszloDev put together [here](https://github.com/LaszloDev/react-admin-firebase/blob/357e873eb219770cec36e757c612e696d72cb0c9/src-demo/src/posts.js) for PR #39.

I made the filter use the line below and it worked!
`filter={{collectionQuery: collection =>
        collection.where("organization.id", "==", organizationId)}}`

In order to stay consistent with the `react-admin` notation, it'd still be great to use notation like this:
`filter={{ 'organization.id': organizationId }}`

As such, I'll leave this issue open.

**Comment by benwinding** [comment]
Thanks @dhstack,

I'll try and implement that in there soon. Using the filter function has greater flexibility, but yes, it's not consistent with `react-admin`.

Cheers,
Ben

**Comment by benwinding** [comment]
Hi @dhstack,

Hopefully this fixes the problem for using nested filters. It accesses the deep object field using dot path notation, like you mentioned above:

```
filter={{ 'organization.id': organizationId }}
```

Should be deployed in version `3.1.15`

Here's the unit-tests for it, let me know if this works for you:

https://github.com/benwinding/react-admin-firebase/blob/cd2a6842e2d8bdb76cb057eba4e06e387944e88e/tests/arrayHelpers.filtering.spec.ts#L92-L102

**Comment by dhstack** [comment]
Works for me! Thanks!

**Comment by dhstack** [comment]
@benwinding This works for filters on one field.

However, it doesn't work for [filtering a list](https://marmelab.com/react-admin/List.html#filtering-the-list) with reference fields.

```
const ProgramFilter = props => (
  <Filter {...props}>
    <ReferenceInput
      label="Organization"
      source="organization.id"
      reference="organizations"
      allowEmpty
    >
      <SelectInput optionText="name" />
    </ReferenceInput>
  </Filter>
);

<List {...props} filters={<ProgramFilter />}>
```

This is the URL that's generated when the above filter is applied:
```
http://localhost:3000/programs?displayedFilters=%7B%22organization.id%22%3Atrue%7D&filter=%7B%22organization%22%3A%7B%22id%22%3A%22ZxaE6XPDbMQWkeADrUYs%22%7D%7D&order=ASC&page=1&perPage=10&sort=id
```
And the [decoded](https://meyerweb.com/eric/tools/dencoder/) version:
```
http://localhost:3000/programs?displayedFilters={"organization.id":true}&filter={"organization":{"id":"ZxaE6XPDbMQWkeADrUYs"}}&order=ASC&page=1&perPage=10&sort=id
```

**Comment by benwinding** [comment]
Oh good, point, I will check for reference input too

**Comment by benwinding** [comment]
Hi @dhstack,

Let me know if commit dadd210 fixes it (version 3.1.17). I've refactored it so filters can either contain nested objects or object path notation, like below:

```
filter={{ 'organization.id': organizationId }}
or
filter={{ 'organization': { id: organizationId} }}
```

Basically there's a function `objectFlatten` that flattens objects into a _normalized_ dot notation. Here's one of the unit tests, for reference. 

https://github.com/benwinding/react-admin-firebase/blob/dadd210b5429bd6be9673cad6225e95812f2a02c/tests/arrayHelpers.filtering.spec.ts#L123-L129

- Also, this might also be related to issue #112, but I'm not sure. 
- And I think we need unit tests for the `GET_MANY_REFERENCE` query, which is currently untested, but I'm not 100% sure how that query works.

Cheers,
Ben

**Comment by dhstack** [comment]
@benwinding That works. As always, thanks for your help!
